src/Native/Unix/System.Native/pal_process.c

```diff
@@ -233,15 +237,70 @@ int32_t SystemNative_ForkAndExecProcess(const char* filename,
+    // is allowed to write to the parent process memory pages.
+
+    // The thing to remember about shared memory vfork() is the documentation is way out of date.
+    // It does the following things:
+    // * creates a new process in the memory space of the calling process.
+    // * blocks the calling thread (not process!) in an uninterruptable sleep
+    // * sets up the process records so the following happen:
+    //   + execve() replaces the memory space in the child and unblocks the parent
+    //   + process exit by any means unblocks the parent
+    //   + ptrace() makes a security demand against the parent process
+    //   + accessing the terminal with read() or write() fail in system-dependent ways
+    // Due to lack of documentation, changing signals in the vfork() child is a bad idea. We don't
+    // do this, but it's worth pointing out.
+
+    // All platforms that provide shared memory vfork() check the parent process's context when
+    // ptrace() is used on the child, thus making setuid() safe to use after vfork(). The fabled vfork()
+    // security hole is the other way around; if a multithreaded host were to execute setuid()
+    // on another thread while a vfork() child is still pending, bad things are possible; however we
+    // do not do that.
+
+    if ((processId = vfork()) == 0) // processId == 0 if this is child process
+#else
+    if ((processId = fork()) == 0) // processId == 0 if this is child process
+#endif
     {
-        success = false;
-        goto done;
-    }
+	// It turns out that child processes depend on their sigmask being set to something sane rather than mask all.
+        // On the other hand, we have to mask all to avoid our own signal handlers running in the child process, writing
+        // to the pipe, and waking up the handling thread in the parent process. This also avoids third-party code getting
+        // equally confused.
+        // Remove all signals, then restore signal mask.
+        sigset_t junk_signal_set;
+        struct sigaction sa_default;
+        struct sigaction sa_old;
+        struct sigaction sa_trash;
+        memset(&sa_default, 0, sizeof(sa_default)); // On some architectures, sa_mask is a struct so assigning zero to it doesn't compile
+        sa_default.sa_handler = SIG_DFL;
+        for (int sig = 1; ; ++sig)
+        {
+            if (sigaction(sig, &sa_default, &sa_old))
+            {
+                if (sig != SIGKILL && sig != SIGSTOP)
+                    break; // No more signals
+            } else {
+                if ((sa_old.sa_flags & SA_SIGINFO)
+                    ? ((void (*)(int))sa_old.sa_sigaction == SIG_IGN || (void (*)(int))sa_old.sa_sigaction == SIG_DFL)
+                    : (sa_old.sa_handler == SIG_IGN || sa_old.sa_handler == SIG_DFL))
+                {
+                    // It has a pre-defined handler -- put it back
+                    sigaction(sig, &sa_old, &sa_trash);
```

Comment:
Most of the signals will fall into this category. So it would be better to query the previous handler fist and then only set it for signals that don't have it set to SIG_IGN or SIG_DFL, instead of setting it and then resetting back.

Reply:
I tried and failed to locate a portable API call for reading the signal handler w/o setting it.

Reply:
When you pass NULL as the `act` parameter and non-NULL as the oldact, it just gets the current value.
sigaction Linux man page says:
> **If act is non-NULL**, the new action for signal signum is installed
    from act.  If oldact is non-NULL, the previous action is saved in
    oldact.

FreeBSD man page says:
> **If act is non-NULL**, it specifies an action ...

OSX man page says:
> **If act is non-zero**, it specifies an action ...